core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java

```diff
@@ -0,0 +1,300 @@
+ */
+@NotThreadSafe
+public class LocalCacheFileInStream extends FileInStream {
+
+  /** Page size in bytes. */
+  protected final long mPageSize;
+
+  private final byte[] mSingleByte = new byte[1];
+  private final Closer mCloser = Closer.create();
+
+  /** Local store to store pages. */
+  private final CacheManager mCacheManager;
+  /** External storage system. */
+  private final FileSystem mExternalFs;
+  /** Path of the file. */
+  private final AlluxioURI mPath;
+  /** File info, fetched from external FS. */
+  private final URIStatus mStatus;
+  private final OpenFilePOptions mOpenOptions;
+
+  /** Stream reading from the external file system, opened once. */
+  private FileInStream mExternalFileInStream;
+  /** Current position of the stream, relative to the start of the file. */
+  private long mPosition = 0;
+  private boolean mClosed = false;
+  private boolean mEOF = false;
+
+  /**
+   * Constructor.
+   *
+   * @param path path of the file
+   * @param options read options
+   * @param externalFs the external file system if a cache miss occurs
+   * @param cacheManager local cache manager
+   */
+  public LocalCacheFileInStream(AlluxioURI path, OpenFilePOptions options, FileSystem externalFs,
+      CacheManager cacheManager) {
+    mPageSize = externalFs.getConf().getBytes(PropertyKey.USER_CLIENT_CACHE_PAGE_SIZE);
+    mPath = path;
+    // Lazy init of status object
+    mStatus = Suppliers.memoize(() -> {
+      try {
+        return externalFs.getStatus(mPath);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }).get();
+    mOpenOptions = options;
+    mExternalFs = externalFs;
+    mCacheManager = cacheManager;
+  }
+
+  @Override
+  public int read() throws IOException {
+    int bytesRead = read(mSingleByte);
+    if (bytesRead == -1) {
+      return -1;
+    }
+    Preconditions.checkState(bytesRead == 1);
+    return BufferUtils.byteToInt(mSingleByte[0]);
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    Preconditions.checkArgument(len >= 0, "length should be non-negative");
+    Preconditions.checkArgument(off >= 0, "offset should be non-negative");
+    if (len == 0) {
+      return 0;
+    }
+    if (mPosition >= mStatus.getLength()) { // at end of file
+      return -1;
+    }
+    int bytesRead = 0;
+    // for each page, check if it is available in the cache
+    while (bytesRead < len && mPosition < mStatus.getLength()) {
+      long currentPage = mPosition / mPageSize;
+      int currentPageOffset = (int) (mPosition % mPageSize);
+      int bytesLeftInPage = (int) Math.min(mPageSize - currentPageOffset, len - bytesRead);
+      PageId pageId = new PageId(mStatus.getFileId(), currentPage);
+      try (ReadableByteChannel cachedData = mCacheManager.get(pageId, currentPageOffset)) {
```

Comment:
how shall we handle IOException here when reading cache? Assume we shall fallback to external stream? Currently the IOE will be escalated.